### ISSUE TITLE
Inconsistent handling of environment option methods for baseUrl

### ISSUE THREAD

**Issue by sethbattin** [cypress-io/cypress#8680]
I really wrote a novel here.  I will tl;dr:

1. Some cypress configuration methods respect ENV for setting `baseUrl`, and some do not.  
2. `baseUrl` is the only reason I would want to use a `cypress.env.json` file, and it doesn't work like that.
3. The past confusion about this behavior is totally understandable, and it persists.

I promise this is in the spirit:
![image](https://user-images.githubusercontent.com/1627760/94372537-30f6c200-00c4-11eb-8117-c5542b465379.png)

---

## Prior discussion

This topic is different from previous issues about this behavior that discuss whether `cypress.env.json` should or should not work for this purpose.  I would like to focus on the fact that [the 4 ENV-based configuration-setting methods](https://docs.cypress.io/guides/guides/environment-variables.html) behave differently from each other.  Perhaps if the consistency problems are resolved, the confusion about this behavior will also be resolved:

https://github.com/cypress-io/cypress/issues/2726
https://github.com/cypress-io/cypress/issues/2499
https://github.com/cypress-io/cypress/issues/909#issuecomment-344363629
https://github.com/cypress-io/cypress/issues/909#issuecomment-370584233
https://github.com/cypress-io/cypress/issues/909#issuecomment-477767447

For reference:
https://docs.cypress.io/guides/guides/environment-variables.html

1. Set in your configuration file
2. Create a `cypress.env.json`
3. Export as `CYPRESS_*`
4. Pass in the CLI as `--env`
5. Plugin (not covered)
6. suite or test config (not covered)

I'm going to leave out option 5 (plugins) because that is a whole other category of running the software.  Although we strictly _could_ write cypress-specific code to resolve the fact that ENV doesn't work very well, ENV is still broken.  Sticking to ENV allows job-running systems to express cypress config in a way that doesn't require perusing the test codebase.  This is important for CI clarity, cross-team understanding, new developers, SREs, etc.

Option 6 is also not relevant since it's not a global setting.

## Current behavior
I'll describe these in the order they are listed in the documentation.  The cypress runner is exemplary for testing this because of its color-coding of the source of configuration settings.

### option 0 (not using anything), default `null`:
```bash
$ cypress open
```
![image](https://user-images.githubusercontent.com/1627760/94367062-ac934780-00a1-11eb-8b47-1cd64926215a.png)

`Cypress.config("baseUrl")` returns null, and all `visit` and `request` calls will error unless they receive fully-qualified URLs.  This means that either test code must contain the domain, or some other config setting must set `baseUrl`.  This method is so atypical that the docs about ENV explicitly suggest setting it.  In the environment setting docs, no less.

### option 1, `cypress.json`:
```bash
$ cat cypress.json
{"baseUrl": "http://json.test.com/"}
$ cypress open
```

![image](https://user-images.githubusercontent.com/1627760/94370109-8080c200-00b3-11eb-93e8-85366d859ef5.png)

Works as expected.  If a _project_ wants this setting, it puts it in a json file and checks it to version control.

### option 2, `cypress.env.json`:

```bash
$ cat cypress.json 
{"baseUrl": "http://json.test.com/"}
$ cat cypress.env.json 
{"baseUrl": "http://baseurl.test.com/", "CYPRESS_BASE_URL": "http://cybaseurl.test.com/"}
$ cypress open
```
![image](https://user-images.githubusercontent.com/1627760/94370135-bc1b8c00-00b3-11eb-9be5-5443e821dfde.png)

Does not set `baseUrl`.  A single developer cannot permanently override their project's setting for their specific task or setup.  :(  They've got other options, so it's not a dealbreaker.

Side note: setting `{env: { baseUrl }}` in `cypress.json` also doesn't work, though it does match `cypres.env.json`
![image](https://user-images.githubusercontent.com/1627760/94370192-174d7e80-00b4-11eb-8e63-7ff47ceb470b.png)

### option 3, actual ENV:

```bash
$ CYPRESS_BASE_URL="http://env.test.com/" cypress open
```

![image](https://user-images.githubusercontent.com/1627760/94370246-84611400-00b4-11eb-9002-64bc456d5ad6.png)

Sets `baseUrl` :+1:   
Does not work like `cypress.env.json` :-1: 
Does not set `env` :question:  

It actually conflicts with [option 3 documentation](https://docs.cypress.io/guides/guides/environment-variables.html#Option-3-CYPRESS):
> Any OS-level environment variable on your machine that starts with either CYPRESS_ or cypress_ will automatically be added to Cypress’ environment variables and made available to you.
> Conflicting values will override values from your configuration file (cypress.json by default) and cypress.env.json files.
> Cypress will strip off the CYPRESS_ when adding your environment variables.

Of these three statements, the only one that is _PARTIALLY_ true is the second one, but only because it sets the `baseUrl`.  But according to [previous explanations](https://github.com/cypress-io/cypress/issues/909#issuecomment-344333167), config and ENV are separate.  That's clearly not true here; setting ENV sets non-environment config and does not set cypress env. :confused: 

### option 4, CLI (two variants):

By the color coding, we see that these variants are the same flavor of setting config.  But these also behave differently from each other.  One sets ENV, and the other sets the thing we meant to set (via ENV).

First variant, `--env`
```bash
$ cypress open --env CYPRESS_BASE_URL="http://flag.test.com"
```
![image](https://user-images.githubusercontent.com/1627760/94370447-a9a25200-00b5-11eb-9e0c-50ead9891493.png)
Sets `env`, but not the setting we really care about. Similar to option 2

Second variant, `--config`
```bash
$ cypress open --config baseUrl="http://config.test.com"
```
![image](https://user-images.githubusercontent.com/1627760/94370821-d48da580-00b7-11eb-8e79-9536b8c366b2.png)

Works for this setting! :tada: 
Doesn't work for another other cypress env, only config. :disappointed: 

### Desired behavior
These are my opinions; I would be happy with any solution that allows migrating between env-setting methods without tripping over the differences.

For option 2: `cypress.env.json` should behave like a local-override of `cypress.json`
1. This is how `.env` files typically behave, and it allows local users to vary from their project default
2. These two files are not redundant; a project file is in version control, and a local file is not.  They have clear use cases.
3. Having to implement env-parsing logic in order to override settings, forbidding simple config files, makes it difficult to migrate from one configuration method to another without a big patch.

For option 3: make it match the docs description, without special handling of `CYPRESS_BASE_URL`
1. Being able to set specific settings via ENV-name-parsing is a little bit arcane, but it works.  
2. If you list specific ENV values that automatically become config values, that's no worse than having to document all the config settings at all.
3. The test runner has its color-coding to resolve any confusion.
4. If ENV requires special parsing logic in order to set `baseUrl`, make that be the case _everywhere_.  Don't allow option 3 to be an exception to that rule (even though it's incredibly convenient, it's confusing).

For option 4: make internal logic consistent.
1. e.g. first set `cypress.env` from `--env`, then set cypress config from `cypress.env`.
2. this allows gradual migrations from simple static settings to more complicated ENV

For option 5: genuinely this isn't a good solution to the problem.  We use ENV because it's explicit and it doesn't complicate the code.  I tried to change my project's default setting for my own local env, and I would have to implement an entirely different bootstrapping process for our suites in order to change one setting.  I just want to inject  `CYPRESS_BASE_URL` into our npm scripts automatically.  They already wrap up `cypress open` or run the node-module version.  Instead i'm going to be mildly annoyed typing longer CLI commands by hand.

I'm sure there are some extreme complications to changing this behavior, but right now it's not good :(

### Test code to reproduce
See above

### Versions
I made all my screenshots with cypress 5.1.0

**Comment by peterjaap** [comment]
I don't understand why this hasn't changed - this is causing utter confusion all over the place.

**Comment by sethbattin** [comment]
@peterjaap this issue was been updated very recently with those labels, actually.  They look like it has been taken into the product roadmap discussion, which is honestly all that I hoped for when I typed this up.

I would expect any real consistency fix to require a big overhaul because the current behavior is the result of a long period of incremental development.  Such is software.

**Comment by lwohlhart** [comment]
+1

**Comment by sethbattin** [comment]
Thanks @cypress-app-bot , I would also love an update!

**Comment by nagash77** [comment]
Hi @sethbattin . thanks for confirming this is still relevant to your usage of Cypress.  I will update with the appropriate labels and move it on to our product team for review.

**Comment by madelashmet** [comment]
thank you for this explanation, @sethbattin ! I was losing my mind trying to figure out why the baseUrl override wasn't taking.

**Comment by lp1641** [comment]
![Image](https://github.com/user-attachments/assets/4bf8f2c1-8b8f-4103-bcc5-f229a7eddee5)
Only works in this way in my case.

**Comment by jorgevds** [comment]
Chiming in to say I could not edit the baseUrl property at all using docker-compose. My setup is a turborepo project where one app has cypress tests, and the only way to run Cypress at all was to run it from the root of the project configured as a turbo command. From there, if I omit cypress.env.json files altogether, I can add the --config flag and point the baseUrl to the host + port of the service dependency in docker-compose, by injecting it into the Dockerfile with a build arg.

**Comment by crystalayman** [comment]
---

## How to Override `baseUrl` in Cypress CI/CD
I've been experiencing a similar problem for the past couple of days and here is what I found

### The Problem

Cypress has inconsistent behavior between configuration methods for `baseUrl`(https://github.com/cypress-io/cypress/issues/8680)):

| Method | Sets `baseUrl`? | Priority |
|--------|-----------------|----------|
| `cypress.config.ts` | ✅ | Lowest |
| `cypress.env.json` | ❌ | N/A |
| `CYPRESS_BASE_URL` env var | ✅ | Medium |
| `--config baseUrl=` (CLI) | ✅ | **Highest** |

Additionally, in **GitHub Actions**, if your URL or environment name matches a secret value, environment variables get **masked** (replaced with `***`), which can break your configuration silently.

---

### Recommended Solution for CI/CD

Use the `config:` option in the Cypress GitHub Action, combined with **GitHub Environment Variables** (not secrets):

```yaml
- name: Cypress run
  uses: cypress-io/github-action@v6
  with:
    config: baseUrl=${{ vars.BASE_URL }}
```

**Why this works:**
1. `--config` has the **highest priority** in Cypress — overrides everything
2. `vars.BASE_URL` is a GitHub **variable** (not a secret) — it won't be masked
3. Avoids the inconsistent `CYPRESS_BASE_URL` environment variable behavior

**Setup:** Add `BASE_URL` as a variable in each GitHub Environment:
- **alpha**: `BASE_URL` = `https://alpha.example.com/`
- **staging**: `BASE_URL` = `https://staging.example.com/`

---

### Overriding `baseUrl` in a Specific Test

#### Option 1: Use a full URL in `cy.visit()`

```typescript
// Ignores baseUrl entirely for this visit
cy.visit('https://different-url.com/some-path');
```

#### Option 2: Override for an entire spec file

```typescript
before(() => {
  Cypress.config('baseUrl', 'https://different-url.com');
});

it('uses the overridden baseUrl', () => {
  cy.visit('/path'); // → https://different-url.com/path
});
```

#### Option 3: Override for a single test

```typescript
it('uses a different baseUrl', { baseUrl: 'https://different-url.com' }, () => {
  cy.visit('/path'); // → https://different-url.com/path
});
```

---

### Priority Summary (Highest to Lowest)

1. **`cy.visit()` with full URL** — in test code
2. **Test/suite config** — `it('test', { baseUrl: '...' }, () => {})`
3. **`--config baseUrl=`** — CLI / GitHub Action `config:` option
4. **`CYPRESS_BASE_URL`** — environment variable
5. **`cypress.config.ts`** — config file

---

### TL;DR

- **CI/CD**: Use `config: baseUrl=${{ vars.BASE_URL }}` with GitHub Environment variables
- **Specific test**: Use `cy.visit('https://full-url.com')` or test-level `{ baseUrl: '...' }` config
- **Avoid**: Relying on `CYPRESS_BASE_URL` env var in CI — it can be masked or behave inconsistently
